oxidation/crates/parsec_api_protocol/src/handshake.rs

```diff
@@ -0,0 +1,668 @@
+    -ballpark_client_early_offset < seconds && seconds < ballpark_client_late_offset
+}
+
+#[derive(Debug, Default, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ApiVersion {
+    pub version: u32,
+    pub revision: u32,
+}
+
+impl PartialOrd for ApiVersion {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ApiVersion {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.version.cmp(&other.version) {
+            Ordering::Equal => self.revision.cmp(&other.revision),
+            order => order,
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(tag = "type")]
+pub enum Answer {
+    #[serde(rename = "AUTHENTICATED")]
+    Authenticated {
+        client_api_version: ApiVersion,
+        organization_id: OrganizationID,
+        device_id: DeviceID,
+        rvk: Box<VerifyKey>,
+        #[serde_as(as = "Bytes")]
+        answer: Vec<u8>,
+    },
+    #[serde(rename = "INVITED")]
+    Invited {
+        client_api_version: ApiVersion,
+        organization_id: OrganizationID,
+        invitation_type: InvitationType,
+        token: InvitationToken,
+    },
+}
+
+impl_dumps_loads!(Answer);
+
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(tag = "type", rename = "signed_answer")]
+pub struct SignedAnswer {
+    #[serde_as(as = "Bytes")]
+    pub answer: [u8; HANDSHAKE_CHALLENGE_SIZE],
+}
+
+impl_dumps_loads!(SignedAnswer);
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub enum HandshakeResult {
+    BadAdminToken,
+    BadIdentity,
+    BadProtocol,
+    Ok,
+    OrganizationExpired,
+    RevokedDevice,
+    RvkMismatch,
+}
+
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(tag = "handshake", rename_all = "snake_case")]
+pub enum Handshake {
+    Challenge {
+        #[serde_as(as = "Bytes")]
+        challenge: [u8; HANDSHAKE_CHALLENGE_SIZE],
+        supported_api_versions: Vec<ApiVersion>,
+        // Those fields have been added to API version 2.4
+        // They are provided to the client in order to allow them to detect whether
+        // their system clock is out of sync and let them close the connection.
+        // They will be missing for older backend so they cannot be strictly required.
+        // TODO: This backward compatibility should be removed once Parsec < 2.4 support is dropped
+        #[serde(default, deserialize_with = "maybe_field::deserialize_some")]
+        ballpark_client_early_offset: Option<f64>,
+        #[serde(default, deserialize_with = "maybe_field::deserialize_some")]
+        ballpark_client_late_offset: Option<f64>,
+        #[serde(default, deserialize_with = "maybe_field::deserialize_some")]
+        backend_timestamp: Option<DateTime>,
+    },
+    Answer(Box<Answer>),
```

Comment:
Why box is needed here ?

Reply:
This was due to the answer size, but it's resolved by boxing rvk field

Reply:
Done

Reply:
> This was due to the answer size

That's an interesting point, do you have a rule of thumb for determining what should be boxed ?

(FYI rvk is a ed25519 public key, so it size is at least 32 bytes, but RustCrypto cache aditional points on the Edward curve so total size is around ~200bytes)

Reply:
It's clippy the goat which gives this advice, the point is that an enum use the same size for all his variants, then it can be a waste of memory if a variant is larger than another

Reply:
good point !